### DISCUSSION TITLE
How to run Blis in parallel

### DISCUSSION THREAD

**Discussion by danilKlyanchin** [coin-or/CHiPPS-BLIS#10] [Q&A]
Hello! 

I want to run the solver in parallel on Mac OS system. I have open-mpi installed. But it looks like sth goes wrong. I'm using the following command to build Blis solver:

```
./coinbrew build Blis --enable-static --disable-shared \
--with-mpi-incdir=/opt/homebrew/Cellar/open-mpi/5.0.0/include \  
--with-mpi-lib="-L/opt/homebrew/Cellar/open-mpi/5.0.0/lib -lmpic++" \   
MPICC=mpicc MPICXX=mpiс++
```

Running ```./blis pathToLp```  gives only non-parallel solver. Okay, but running   ```mpirun -n 2 blis pathToLp``` launch a non-parallel solver 2 times simultaneously and independently. I guess, that is not how Blis was developed to work and sth is wrong. 

I realized, that the code has an essential flag COIN_HAS_MPI which should be set to true. But I have this flag unset and I do not know how to set in correctly. 
Can anybody give me an advice how to run Blis correctly in parallel?

**Comment by tkralphs** [comment]
What you are doing should indeed be working, provided that MPI is successfully found and linked. It's possible that is not happening and so the build is just happening in sequential mode without failing. If you can attach `config.log` from the Blid build directory, I can try to diagnose, but since I haven't run Blis in parallel myself for a while, I'll also try to make sure it still builds. What version are you building? If you are building `master` or `0.95`, those use a newer version of the build system and it may be that some recent changes there broke something. Building the `0.94` would be safer (there aren't many substantive differences between the two versions anyway). You should not need to manually define any symbols.

**Reply by danilKlyanchin** [reply]
Yes, I have used master version of Blis. Here is the ```config.log``` file.
[config.log](https://github.com/coin-or/CHiPPS-BLIS/files/14098361/config.log)


I also tried the 0.94 version as you said, the results are:
1) Running the command ```./coinbrew build Blis --enable-static --disable-shared --with-mpi-incdir=/opt/homebrew/Cellar/open-mpi/5.0.0/include --with-mpi-lib="-L/opt/homebrew/Cellar/open-mpi/5.0.0/lib -lmpic++" MPICC=mpicc MPICXX=mpiс++```
I get the warning ```configure: error: Cannot find symbol(s) MPI_Irecv with MPI``` in the Alps building time. May be the build process is correct but MPI was installed incorrectly... 

2) Running the command  ```./coinbrew build Blis --enable-static --disable-shared --with-mpi-cflags=/opt/homebrew/Cellar/open-mpi/5.0.0/include --with-mpi-lflags="-L/opt/homebrew/Cellar/open-mpi/5.0.0/lib -lmpic++" MPICC=mpicc MPICXX=mpic++``` I have no warnings, but the behaviour is the same as I mentioned 5 days ago about master version.

**Reply by tkralphs** [reply]
I was able to eventually build successfully with Open MPI using
```
coinbrew build Blis --enable-debug --enable-static --disable-shared --with-mpi-lib="-L/usr/lib/x86_64-linux-gnu/openmpi/lib -lmpi" --with-mpi-incdir=/usr/lib/x86_64-linux-gnu/openmpi/include MPICC=mpicc MPICXX=mpic++ -j 4 -b build-blis-0.92-mpi-debug -p
```
I say "eventually" because there was a small fix needed to allow the build to go through. I then found that there was at least one and probably two small bugs introduced in the last few years that broke the MPI version. One of these is easy to fix, the other is not completely obvious to me. So Blis `0.94` doesn't currently work out of the box. 

I was able to get a version that seems to be fully working by going back to `0.92`. For this, I checked out `0.94` with `coinbrew` and then manually switched the versions Alps, Bcps, and Blis to `1.3`, `0.92` and `0.92, respectively. (You can't really use coinbrew because older versions of the other dependencies won't buil with newer C++ compilers. This gave me a working version of Blis that I could successfully run in parallel with MPI. 

Looking at the `config.log` you attached, it seems to be for a build in the `master` branch, since the `configure` script being run is in the root directory, while for older version of Blis, this would instead be in the `Blis/` subdirectory. As I suspected, the `configure` script is simply not finding the MPI compiler.  In my `config.log`, there is a line that reads
```
configure:3245: Will use MPI C compiler mpicc
```
You don't have such a line. There could be two reasons. In the `master` branch, it is entirely possible that support for the MPI compiler was accidentally broken or removed, since this is a relatively new version of the build system and MPI is not much used anymore. For `0.94`, this wouldn't be the reason, though, so it must be that `mpicc/mpic++` wasn't found in your path or something to that effect. If you can provide me with the `config.log` for that build, we can diagnose it. 

But in any case, version `0.94` won't work until I track down the bug I described above. Version `0.92` should be fine, though. There haven't been many substantive changes in the code since then. If you can figure out the reason why `mpicc` is not used in the build, then we should be able to get `0.92` up and running.

**Comment by danilKlyanchin** [comment]
Hi there again! 
Recently I came back to the problem of running Blis in parallel. After some tries of compiling and with the help of your comments about 0.94 and 0.92 versions of Blis I successfully run Blis in parallel!
I decided to describe how I compiled Blis, step by step, maybe I still don't understand something and someone will help me, or maybe it will help someone.

So:
1. As you have said, I installed Blis@0.94 version using coinbrew
```
./coinbrew fetch Blis@0.94
```
2. Then I manually switched the versions of Alps, Bcps and Blis to 1.3, 0.92 and 0.92 respectively
```
rm -rf Alps Bcps Blis
./coinbrew fetch Alps@1.3 --skip-dependencies
./coinbrew fetch Bcps@0.92 --skip-dependencies
./coinbrew fetch Blis@0.92 --skip-dependencies
```
3. Then I added some additional flags
```
export CC=mpiicc
export CXX=mpiicpc
export CXXFLAGS="-std=c++17 -DCOIN_HAS_MPI"
export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:/opt/intelmpi/mpi/latest/lib/pkgconfig
```
4. Compile command
```
./coinbrew build Blis \
  --enable-static --disable-shared \
  --with-mpi-lib="-L/opt/intelmpi/mpi/latest/lib -lmpi" \
  --with-mpi-incdir=/opt/intelmpi/mpi/latest/include \
  MPICC=mpiicc MPICXX=mpiicpc
```

And here the compilation succeed. But the question is, is it ok to use `-DCOIN_HAS_MPI` additional flag? Because without this flag compilation fails... The question arrises as I didn't see anything in the documentations about using this flag.

The last but not least question about parallel Blis is a message `WARNING: Subtree size is larger than message buffer size, will split it` which occurs quite often and, moreover, invokes a death of all processes. Could it be fixed by proposing some additional flags or what is it really trying to tell me? :) 
The original .lp problem which cases the warning is air05 (here is the link https://miplib.zib.de/instance_details_air05.html).
